packages/wonder-stuff-core/src/error-info.js

```diff
@@ -0,0 +1,200 @@
+// @flow
+import {buildCausedByMessage} from "./build-caused-by-message.js";
+
+/**
+ * Normalized error message and stack.
+ */
+export class ErrorInfo {
+    +_message: string;
+    +_stackFrames: $ReadOnlyArray<string>;
+
+    constructor(message: string, stackFrames: $ReadOnlyArray<string>) {
+        this._message = message;
+        // Copy the stack array just to be sure our internal state won't
+        // mutate outside our control.
+        this._stackFrames = [...stackFrames];
+    }
+
+    /**
+     * The normalized error message associated with this stack trace.
+     */
+    get message(): string {
+        return this._message;
+    }
+
+    /**
+     * An array representing the stack frames of the error.
+     *
+     * Index 0 being the top of the stack.
+     */
+    get stack(): $ReadOnlyArray<string> {
+        // Copy the stack array just to be sure our internal state won't
+        // mutate outside our control.
+        return [...this._stackFrames];
+    }
+
+    /**
+     * Get the stack trace as a string.
+     */
+    toString(): string {
+        // Now construct the string value.
+        return `${this._message}\n${this._stackFrames.join("\n")}`;
+    }
+
+    /**
+     * Combine the stack of the error that caused this one.
+     *
+     * This combines the stack of another error into this one, creating a new
+     * stack trace that incorporates the frames of both errors to tell a deeper
+     * story of what happened.
+     *
+     * For example, some work is done and error A gets thrown, this is caught
+     * and ultimately wrapped by error B, which then gets thrown. By combining
+     * the stack traces of both error A and B we can get a better idea of what
+     * happened, which can be especially useful in asynchronous code.
+     *
+     * Stack traces are combined such that common frames are deduped, and then
+     * at the point of the original throw of error A (i.e. error A's top stack
+     * frame), the deduped trace of error B is inserted to show what happened
+     * after the error was caught.
+     *
+     * So, given:
+     *
+     * cause = ErrorA, with stack,
+     *    Error: A
+     *      stackA2
+     *      stackA1
+     *      stackShared2
+     *      stackShared1
+     *
+     * and consequence = ErrorB, with stack,
+     *    Error: B
+     *       stackB2
+     *       stackB1
+     *       stackShared2
+     *       stackShared1
+     *
+     * This method would produce the equivalent of:
+     *    Error: B
+     *        caused by
+     *            Error: A
+     *      stackA2
+     *      stackA1
+     *      stackB2
+     *      stackB1
+     *      stackShared2
+     *      stackShared1
```

Comment:
oooooh, I understand now why this method is also doing the message-combining too! It's because the calls stacks of errors start with the error message on the first frame!

Hmmm... I think they way Java does it is more like
```
Error: B 
   frame
   frame
caused by
Error A
    frame
    frame
```

I don't really know if one way is better than the other or easier than the other... especially if we go n-deep on error nesting...

But this and the fact that `toString()` is the error message only, on mobile, I'll think more on this.

Anyway, this is only the first diff! Feel free to land and increment changes! :-)

Reply:
Yeah, I'm looking to construct a stack that matches JavaScript platforms, which tend to expect a message on the first line, but then stack frames for the rest. And this is only used in the case of adding a cause to an error, so it doesn't affect the other possible things we will be building messages for.

Reply:
Even with nesting, this will all work as the same rules will apply and, just as with the mobile code now, the combining is done during error construction and the same rules still apply.